### ISSUE TITLE
Code

### ISSUE THREAD

**Issue by bhagwatsolunke** [bhagwatsolunke/dl#2]
Pract 3 
*************
import numpy as np
import pandas as pd
import random
import tensorflow as tf
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten,Conv2D,Dense,MaxPooling2D
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.datasets import mnist

(X_train,y_train),(X_test , y_test) = mnist.load_data()
print(X_train.shape)

print(X_train.shape)

X_train[0].min() , X_train[0].max()

#image reduction in range of 0(black) to 255(white)
X_train = (X_train - 0.0) / (255.0 - 0.0)
X_test = (X_test - 0.0 ) / (255.0 - 0.0)
X_train[0].min(),X_train[0].max()

def plot_digit(image,digit,plt,i):
  plt.subplot(4,5,i+1)
  plt.imshow(image , cmap=plt.get_cmap('gray'))
  plt.title(f"Digit: {digit}")
  plt.xticks([])
  plt.yticks([])
plt.figure(figsize=(16,10))
for i in range(20):
  plot_digit(X_train[i],y_train[i],plt,i)
plt.show()

#We added as channel making the img form 2D to 3D as most DL model work with that
X_train = X_train.reshape((X_train.shape + (1,)))
X_test = X_test.reshape((X_test.shape + (1,)))

model = Sequential([
    #First layer : convolutional layer ->32 filters (3,3) the filter size is 3*3, input shape model expects a img of 28*28 pixels and single channels(grayscale)
    Conv2D(32,(3,3) , activation="relu", input_shape=(28,28,1)),
    #Adds max pooling to reduce the sptail dimensions of the previous layer'output
    MaxPooling2D((2,2)),
    #Converts the previous layers 2D output and converts is in 1D
    Flatten(),
    Dense(100,activation="relu"),
    #Last layer the ouptut layer, Softmax function computes the probablity over the 10 classes
    Dense(10,activation="softmax")
])

#Create a instance of SGD(stochastic Gradient Descent)
#learing rate used to control the step size to update the model weight
#moentum helps accelarate convergence
optimizer = SGD(learning_rate=0.01, momentum=0.9)
model.compile(
    optimizer = optimizer,
    #what to consider for parameters while training
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

model.summary()

import random
plt.figure(figsize=(16,10))
for i in range (20):
  image = random.choice(X_test).squeeze()
  digit = np.argmax(model.predict(image.reshape((1,28,28,1)))[0],axis=-1)
  plot_digit(image,digit,plt,i)
plt.show()

n = random.randint(0,9999)
plt.imshow(X_test[n])
plt.show()

predicted_value = model.predict(X_test)
print("HandWritten number in the image is=%d" %np.argmax(predicted_value[n]))

score = model.evaluate(X_test,y_test,verbose=0)
print('Test Loss:', score[0])
print('Test accuracy:', score[0])

#fit method return training history
model_log = model.fit(
    X_train,             # Training data
    y_train,             # Training labels
    epochs=10,           # Number of training epochs (iterations)
    batch_size=15,       # Batch size for training
    verbose=1,           # Verbosity level (1 for progress updates, 0 for silent)
    validation_data=(X_train, y_train)  # Validation data and labels
)

**************
Pract 5
**************
import torch
import torch.nn as nn
import torch.optim as optim

sentences = [
    "I don't love deep learing",
    "I don't enjoy working with neural networks",
    "Pytorch is maybe great, i don't know how to use it"
]

#Vocabulary Creation
word_to_idx = {}
idx = 0
for sentence in sentences:
  words = sentence.split()
  for word in words:
    if word not in word_to_idx:
      word_to_idx[word] = idx
      idx+=1
#Inverse dictionary to map indicates back to words
idx_to_word = {idx:word for word,idx in word_to_idx.items()}

context_window = 2

#Genrate training data
def generate_cbow_training_data(sentences, word_to_idx, context_window):
    data = []
    for sentence in sentences:
        words = sentence.split()
        for i in range(len(words)):
          target_word = words[i]
          context = []
          for j in range(max(0,i-context_window), min(i+context_window+1, len(words))):
            if j!= i:
              context.append(word_to_idx[words[j]])
          for word in context:
            data.append((word,word_to_idx[target_word]))

    return data

training_data = generate_cbow_training_data(sentences,word_to_idx,context_window)

#Train the model
#Creating a CBOW Model
class CBOW(nn.Module):
  def __init__(self,vocab_size,embedding_dim):
    super(CBOW,self).__init__()
    self.embedding = nn.Embedding(vocab_size,embedding_dim)
    self.linear = nn.Linear(embedding_dim,vocab_size)

  def forward(self,context):
    embedded = self.embedding(context).sum(dim=1)
    out = self.linear(embedded)
    return out

#Defining the CBOW model, loss function, and optimizer
# Define the embedding dimension
embedding_dim = 1  # You can choose the desired dimension

# Create the CBOW model, loss function, and optimizer
cbow_model = CBOW(len(word_to_idx), embedding_dim)
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(cbow_model.parameters(), lr=0.1)

#Training Loop
for context, target in training_data:
    context = torch.tensor([context], dtype=torch.long)  # Wrap context in a list to make it 2D
    target = torch.tensor(target, dtype=torch.long)

    optimizer.zero_grad()
    outputs = cbow_model(context)

    # Loss for each word in the context
    loss = 0
    for i in range(context.shape[1]):
        loss += criterion(outputs, target.view(-1))  # Compute loss for each word

    loss.backward()
    optimizer.step()

****************
Pract 6
****************
from tensorflow.keras.utils import load_img
from tensorflow.keras.utils import img_to_array
from keras.applications.vgg16 import preprocess_input
from keras.applications.vgg16 import decode_predictions
from keras.applications.vgg16 import VGG16

#loads the image and resizes it to a target size of 224*224
image = load_img("cat.jpeg",target_size=(224,224))
#Convert the image in numpy array
image = img_to_array(image)
image = image.reshape((1,image.shape[0],image.shape[1],image.shape[2]))
#Function is used to preprocess the image according to the specific requirements of the VGG16 models
#preprocessing include scaling,substaraction and other
image = preprocess_input(image)
#Create the instance of VGG16 model
model = VGG16
# yhat will contain the prediction which is the probablity distribution over diff classes
yhat = model.predict(image)
label = decode_predictions(yhat)
label = label[0][0]
print('%s(%.2f%%)' % (label[1],label[2]*100))
